## Loading the standard libraries

In [1]:
import pandas as pd
import numpy as np

In [33]:
dic = {'name' : ['Anil', 'Sunil', 'Ashok', 'John', 'Sam', 'Peter'],
       'Age' : [45, np.nan, 40, 38, np.nan, 65],
       'qualification' : [np.nan, 'BE', 'ME', 'BCA', 'MCA', np.nan],
       'Salary' : [35000, 75000, np.nan, 48000, 76000, np.nan]
}
df  = pd.DataFrame(dic)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           NaN      NaN

In [3]:
df.shape

(6, 4)

## Observations

- NaN full form is Not a Number
- Missing values in the data are respresented by NaN

## To check data having missing value or not?

In [4]:
df.isnull()

name    Age  qualification  Salary
0  False  False           True   False
1  False   True          False   False
2  False  False          False    True
3  False  False          False   False
4  False   True          False   False
5  False  False           True    True

## To check cols having missing values in the data

In [5]:
df.isnull().any()

name             False
Age               True
qualification     True
Salary            True
dtype: bool

In [6]:
data = pd.read_csv('titanic.csv')
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [7]:
data.shape

(891, 12)

In [8]:
data.isnull()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False   True  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked  
0    False   True     False  
1    False  False     False  
2    False   True     False  
3    False  False     False  
4    False   True     False  
..     ...    ...       ...  
886  False   True     False  
887  False  False     False  
888  False   True     False  
889  False  False     False  
890  False   True     False  

[891 rows x 12 columns]

In [9]:
data.isnull().any()

PassengerId    False
Survived       False
Pclass         False
Name           False
Sex            False
Age             True
SibSp          False
Parch          False
Ticket         False
Fare           False
Cabin           True
Embarked        True
dtype: bool

## Observation 

- Only 3 columns, Age, Cabin, Embarked are having missing value in the data

# How to count the total missing value present in each column

In [11]:
df.isnull()

name    Age  qualification  Salary
0  False  False           True   False
1  False   True          False   False
2  False  False          False    True
3  False  False          False   False
4  False   True          False   False
5  False  False           True    True

In [12]:
df.isnull().sum()

name             0
Age              2
qualification    2
Salary           2
dtype: int64

In [13]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

## To check the percentage of missing values in each column

In [14]:
data.isnull().sum() / len(data) * 100

PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
Age            19.865320
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
Cabin          77.104377
Embarked        0.224467
dtype: float64

In [15]:
len(data)

891

## Insights :

1. Cabin is having 77.10 % missing values 
2. Embarked is having 0.22 % missing values
3. Age is having 19.86% missing values


In [16]:
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           NaN      NaN

## Missing Value Treatment(Imputation) techniques


### Dropping the missing values

In [17]:
df.dropna()

name   Age qualification   Salary
3  John  38.0           BCA  48000.0

## As per standard it is advisable to drop a column having more than 30 % missing values 

- In the titanic data, Cabin is having 77% missing values hence drop the entire the column Cabin column from the data

In [18]:
data = data.drop('Cabin', axis = 1)
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

In [19]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Embarked         2
dtype: int64

## When using dropna() techniques ensure that, the particular column is having less than 1% missing values

- In the titanic data, Embarked is having less than 1% missing values hence use the dropna technique

In [20]:
data = data.dropna()
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

In [21]:
data.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

## Load the data

In [38]:
data = pd.read_csv('titanic.csv')
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [39]:
data = data.drop('Cabin', axis = 1)
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

## Missing Value Imputation techniques

### Numerical Column Missing value Imputation:

1. Mean
2. Median
3. Backward Fill
4. Forward fill
5. Constant Value

### Categorical Column Missing value imputation:

1. Mode Imputation
2. Backward Fill
3. Forward Fill
4. Constant Value

In [24]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Embarked         2
dtype: int64

## Constant value imputation on Age:

In [25]:
data['Age'].fillna(30, inplace = True)
data.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       2
dtype: int64

## Mean value imputation on Age column

In [28]:
data['Age'].mean()

29.69911764705882

In [29]:
data['Age'].fillna(data['Age'].mean(), inplace = True)
data.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       2
dtype: int64

In [30]:
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           NaN      NaN

In [31]:
df['Age'].mean()

47.0

In [35]:
df['Age'].fillna(df['Age'].mean(), inplace = True)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil  47.0            BE  75000.0
2  Ashok  40.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam  47.0           MCA  76000.0
5  Peter  65.0           NaN      NaN

## Median Imputation 

In [36]:
df['Salary'].median()

61500.0

In [37]:
df['Salary'].fillna(df['Salary'].median(), inplace = True)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil  47.0            BE  75000.0
2  Ashok  40.0            ME  61500.0
3   John  38.0           BCA  48000.0
4    Sam  47.0           MCA  76000.0
5  Peter  65.0           NaN  61500.0

In [40]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Embarked         2
dtype: int64

In [41]:
data['Age'].fillna(data['Age'].median(), inplace = True)
data.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       2
dtype: int64

## Backward Fill imputation

In [42]:
dic = {'name' : ['Anil', 'Sunil', 'Ashok', 'John', 'Sam', 'Peter'],
       'Age' : [45, np.nan, 40, 38, np.nan, 65],
       'qualification' : [np.nan, 'BE', 'ME', 'BCA', 'MCA', np.nan],
       'Salary' : [35000, 75000, np.nan, 48000, 76000, np.nan]
}
df  = pd.DataFrame(dic)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           NaN      NaN

In [44]:
df['Age'].fillna(method = 'bfill', inplace = True)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil  40.0            BE  75000.0
2  Ashok  40.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam  65.0           MCA  76000.0
5  Peter  65.0           NaN      NaN

In [45]:
df['Salary'].fillna(method = 'bfill', inplace = True)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil  40.0            BE  75000.0
2  Ashok  40.0            ME  48000.0
3   John  38.0           BCA  48000.0
4    Sam  65.0           MCA  76000.0
5  Peter  65.0           NaN      NaN

In [46]:
dic = {'name' : ['Anil', 'Sunil', 'Ashok', 'John', 'Sam', 'Peter'],
       'Age' : [45, np.nan, np.nan, 38, np.nan, 65],
       'qualification' : [np.nan, 'BE', 'ME', 'BCA', 'MCA', np.nan],
       'Salary' : [35000, 75000, np.nan, 48000, 76000, np.nan]
}
df  = pd.DataFrame(dic)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok   NaN            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           NaN      NaN

In [47]:
df['Age'].fillna(method = 'bfill', inplace = True)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil  38.0            BE  75000.0
2  Ashok  38.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam  65.0           MCA  76000.0
5  Peter  65.0           NaN      NaN

## Forward fill imputation

In [48]:
dic = {'name' : ['Anil', 'Sunil', 'Ashok', 'John', 'Sam', 'Peter'],
       'Age' : [45, np.nan, 40, 38, np.nan, 65],
       'qualification' : [np.nan, 'BE', 'ME', 'BCA', 'MCA', np.nan],
       'Salary' : [35000, 75000, np.nan, 48000, 76000, np.nan]
}
df  = pd.DataFrame(dic)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           NaN      NaN

In [49]:
df['Salary'].fillna(method = 'ffill', inplace = True)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            ME  75000.0
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           NaN  76000.0

## Missing Value imputation for Categorical data

In [50]:
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            ME  75000.0
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           NaN  76000.0

### Constant Value Imputation

In [51]:
df['qualification'].fillna('BTech', inplace = True)
df

name   Age qualification   Salary
0   Anil  45.0         BTech  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            ME  75000.0
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0         BTech  76000.0

### Backward fill

In [52]:
dic = {'name' : ['Anil', 'Sunil', 'Ashok', 'John', 'Sam', 'Peter'],
       'Age' : [45, np.nan, 40, 38, np.nan, 65],
       'qualification' : [np.nan, 'BE', 'ME', 'BCA', 'MCA', np.nan],
       'Salary' : [35000, 75000, np.nan, 48000, 76000, np.nan]
}
df  = pd.DataFrame(dic)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           NaN      NaN

In [53]:
df['qualification'].fillna(method = 'bfill', inplace = True)
df

name   Age qualification   Salary
0   Anil  45.0            BE  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           NaN      NaN

## Forward fill

In [54]:
dic = {'name' : ['Anil', 'Sunil', 'Ashok', 'John', 'Sam', 'Peter'],
       'Age' : [45, np.nan, 40, 38, np.nan, 65],
       'qualification' : [np.nan, 'BE', 'ME', 'BCA', 'MCA', np.nan],
       'Salary' : [35000, 75000, np.nan, 48000, 76000, np.nan]
}
df  = pd.DataFrame(dic)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           NaN      NaN

In [55]:
df['qualification'].fillna(method = 'ffill', inplace = True)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           MCA      NaN

## Mode Imputation

In [56]:
data = pd.read_csv('loan_data_set.csv')
data.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [57]:
data.shape

(614, 13)

In [58]:
data.isnull().sum()

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [59]:
data['Gender'].value_counts()

Male      489
Female    112
Name: Gender, dtype: int64

## Mode is defined as the most repetative value in the column.

- For the loan data, Male is the mode for the column

In [64]:
data['Gender'].mode()

0    Male
Name: Gender, dtype: object

In [62]:
data['Gender'].mode()[0]

'Male'

In [63]:
data['Gender'].fillna(data['Gender'].mode()[0], inplace = True)
data.isnull().sum()

Loan_ID               0
Gender                0
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

## Mode impuatation on df

In [70]:
dic = {'name' : ['Anil', 'Sunil', 'Ashok', 'John', 'Sam', 'Peter'],
       'Age' : [45, np.nan, 40, 38, np.nan, 65],
       'qualification' : [np.nan, 'BE', 'BE', 'MCA', 'MCA', np.nan],
       'Salary' : [35000, 75000, np.nan, 48000, 76000, np.nan]
}
df  = pd.DataFrame(dic)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            BE      NaN
3   John  38.0           MCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           NaN      NaN

In [68]:
df['qualification'].mode()

0     BE
1    MCA
Name: qualification, dtype: object

In [71]:
df['qualification'].fillna(df['qualification'].mode()[0], inplace = True)
df

name   Age qualification   Salary
0   Anil  45.0            BE  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            BE      NaN
3   John  38.0           MCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0            BE      NaN

## Missing Value imputation using Sklearn library(SimpleImputer Package)

In [73]:
from sklearn.impute import SimpleImputer
si = SimpleImputer()

In [74]:
dic = {'name' : ['Anil', 'Sunil', 'Ashok', 'John', 'Sam', 'Peter'],
       'Age' : [45, np.nan, 40, 38, np.nan, 65],
       'qualification' : [np.nan, 'BE', 'ME', 'BCA', 'MCA', np.nan],
       'Salary' : [35000, 75000, np.nan, 48000, 76000, np.nan]
}
df  = pd.DataFrame(dic)
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil   NaN            BE  75000.0
2  Ashok  40.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam   NaN           MCA  76000.0
5  Peter  65.0           NaN      NaN

## Mean Imputation on Age using SimpleImputer

In [76]:
from sklearn.impute import SimpleImputer
si = SimpleImputer(strategy = 'mean')

df['Age'] = si.fit_transform(df[['Age']])

In [77]:
df['Age']

0    45.0
1    47.0
2    40.0
3    38.0
4    47.0
5    65.0
Name: Age, dtype: float64

In [78]:
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil  47.0            BE  75000.0
2  Ashok  40.0            ME      NaN
3   John  38.0           BCA  48000.0
4    Sam  47.0           MCA  76000.0
5  Peter  65.0           NaN      NaN

## Median Imputation using sklearn Simple Imputer package

In [79]:
from sklearn.impute import SimpleImputer
si = SimpleImputer(strategy = 'median')

df['Salary'] = si.fit_transform(df[['Salary']])

In [80]:
df

name   Age qualification   Salary
0   Anil  45.0           NaN  35000.0
1  Sunil  47.0            BE  75000.0
2  Ashok  40.0            ME  61500.0
3   John  38.0           BCA  48000.0
4    Sam  47.0           MCA  76000.0
5  Peter  65.0           NaN  61500.0

## Mode imputation using sklearn SimpleImputer package

- strategy = most_frequent represents mode of the column

In [82]:
from sklearn.impute import SimpleImputer
si = SimpleImputer(strategy = 'most_frequent')

df['qualification'] = si.fit_transform(df[['qualification']])

In [83]:
df

name   Age qualification   Salary
0   Anil  45.0           BCA  35000.0
1  Sunil  47.0            BE  75000.0
2  Ashok  40.0            ME  61500.0
3   John  38.0           BCA  48000.0
4    Sam  47.0           MCA  76000.0
5  Peter  65.0           BCA  61500.0